
// THIS FILE IS AUTOMATICALLY GENERATED FROM ndap-spec.org using org-babel-mode
// Do not edit it by hand!  Edit the original and re-generate
// Instructions are found in ndap-spec.org!
//
// Contact Eric Anderson <andersoe@cs.cmu.edu> with questions.

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Messages][Messages:1]]

message NetDescriptor {
        optional bytes GUID = 1;
        optional LayerTwoIdentifier l2_id = 2;
        required AuthCapStruct auth_cap = 3;
        optional AuthCapSharedInfo ac_shared = 4;
}

// Messages:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Layer%20Two%20Network][Layer\ Two\ Network:1]]

message LayerTwoIdentifier {
        
        // Identifies the type of Layer 2 Network referred to by this Descriptor        
        enum LayerTwoType {
                ETHERNET = 0;
                WIFI = 1;
                DSRC = 2;
        }
        required LayerTwoType l2_type =1;

        // Meaning of l2id is left up to L2-specific implementation.
        optional bytes l2_net_id = 2;
}

// Layer\ Two\ Network:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Authorization%20and%20Capabilities][Authorization\ and\ Capabilities:1]]

message ACNode {
        optional JoinAuthNode ja = 1;
        optional XIPNode xip = 2;
        // TO BE EXTENDED
}

// Authorization\ and\ Capabilities:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Join/Auth/Config%20Protocol][Join/Auth/Config\ Protocol:1]]

message JoinAuthNode {
        required uint32 version =1;
        required uint32 aaa_provider=2;
        optional uint32 context=3;              // Who knows if this will be used
}

// Join/Auth/Config\ Protocol:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*XIP][XIP:1]]

message XIPNode {
        required bool is_global = 1;
        optional bytes NetworkId = 2;
        optional bytes HostId = 3;
}

// XIP:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Graph%20Structure][Graph\ Structure:1]]

message AuthCapStruct {
        repeated ACNode nodes = 1;
        repeated ACEdge edges = 2;
}

// Graph\ Structure:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Graph%20Structure][Graph\ Structure:1]]

message ACEdge {
        // Indexed base 1
        required uint32 from_node =1;
        required uint32 to_node = 2;
}

// Graph\ Structure:1 ends here

// [[file:~/work/git/xia-vehicular/veh/ndap-spec.orgmk::*Shared%20Information][Shared\ Information:1]]

message AuthCapSharedInfo {
    message JoinAuthExtraInfo {
        message PublicKey {
            enum CipherSuite {
                NACL_curve25519xsalsa20poly1305 = 17;  //The standard (only?) suite recommended by NaCl.
            }
            required CipherSuite cipher_suite = 1;
            required bytes the_key = 2;
        }
        optional PublicKey gateway_ephemeral_pubkey = 1;
        optional bytes gateway_nonce = 2;
    }
    optional JoinAuthExtraInfo ja = 1;
}

// Shared\ Information:1 ends here
